go/kbfs/data/node_obfuscator.go

```diff
@@ -0,0 +1,145 @@
+	return fmt.Sprintf("{private %d bytes}", len(nos))
+}
+
+// NodeObfuscator takes a secret, and uses it to create obfuscate
+// strings based on BIP-0039 dictionary words.  It remembers previous
+// obfuscations, and if there happens to be a collision on the
+// obfuscated text for a new plaintext string, it appends a numeric
+// suffix to the new obfuscation.
+type NodeObfuscator struct {
+	secret NodeObfuscatorSecret
+
+	lock     sync.RWMutex
+	obsCache map[string]string
+	usedObs  map[string]bool
+}
+
+var _ Obfuscator = (*NodeObfuscator)(nil)
+
+// NewNodeObfuscator creates a new `NodeObfuscator` instance.
+func NewNodeObfuscator(secret NodeObfuscatorSecret) *NodeObfuscator {
+	return &NodeObfuscator{
+		secret: secret,
+		// Don't initialize caches yet, for memory reasons, in case
+		// they're never used.
+	}
+}
+
+func (no *NodeObfuscator) checkCacheIfExistsLocked(
+	plaintext string) (string, bool) {
+	if no.obsCache == nil {
+		return "", false
+	}
+
+	obs, ok := no.obsCache[plaintext]
+	return obs, ok
+}
+
+func (no *NodeObfuscator) checkCacheIfExists(plaintext string) (string, bool) {
+	no.lock.RLock()
+	defer no.lock.RUnlock()
+	return no.checkCacheIfExistsLocked(plaintext)
+}
+
+func (no *NodeObfuscator) obfuscateWithHasher(
+	plaintext string, hasher func(string) []byte) string {
+	obs, ok := no.checkCacheIfExists(plaintext)
+	if ok {
+		return obs
+	}
+
+	no.lock.Lock()
+	defer no.lock.Unlock()
+	// See if it's been cached since we last released the lock.
+	obs, ok = no.checkCacheIfExistsLocked(plaintext)
+	if ok {
+		return obs
+	}
+
+	if no.obsCache == nil {
+		no.obsCache = make(map[string]string)
+		no.usedObs = make(map[string]bool)
+	}
+
+	// HMAC the plaintext with the secret, to get a hash.
+	buf := hasher(plaintext)
+
+	// Look up two words based on the first three bytes of the mac.
+	// (Each word takes 11 bits to lookup a unique word among the 2048
+	// secret words.)
+	first := int(buf[0])
+	// First 8 bits are from buf[0].
+	first <<= 3
+	// Last 3 bits are the first 3 bits of buf[1].
+	const first3 = 0x7 << 5
+	first |= ((int(buf[1]) & first3) >> 5)
+	if float64(first) >= math.Exp2(11) {
+		panic(fmt.Sprintf("Generated %d from 11 bits", first))
+	}
+
+	// First 5 bits are the last 5 bits of buf[1].
+	const last5 = 0x1f
+	second := int(buf[1]) & last5
+	second <<= 6
+	// Last 6 bits are the first 6 bits of buf[2].
+	const first6 = 0x3f << 2
+	second |= ((int(buf[2]) & first6) >> 2)
+	if float64(second) >= math.Exp2(11) {
+		panic(fmt.Sprintf("Generated %d from 11 bits", second))
+	}
+
+	firstWord := libkb.SecWord(first)
+	secondWord := libkb.SecWord(second)
+	obs = strings.Join([]string{firstWord, secondWord}, separator)
+
+	suffix := 1
+	for no.usedObs[obs] {
+		suffix++
+		obs = strings.Join(
+			[]string{firstWord, secondWord, strconv.Itoa(suffix)}, separator)
+	}
+
+	no.obsCache[plaintext] = obs
+	no.usedObs[obs] = true
+	return obs
+}
+
+func (no *NodeObfuscator) defaultHash(plaintext string) []byte {
```

Comment:
@maxtaco, @mlsteele: this function will be used to generated hashes for obfuscating a filename in the KBFS logs.  The `secret` here will be a sha256 hash of the TLF's 1st key generation, and `plaintext` will be the name of the file or directory being obfuscated.  Then, the first 3 bytes of the resulting hmac returned by this function will be used to index into the bip-0039 wordlist to get 2 words that can replace the plaintext word.  (Yes, there will be collisions, and that's fine.)  This is an expanded version of an idea @mlsteele wrote up in KBFS-962.

I'd like a sanity check on this to make sure it won't leak any sensitive data, either about the TLF's key (crucial), or the `plaintext` being obfuscated (ideal, but less crucial).  Thanks!

Reply:
I can't think of any way this can be used to leak data either.

Reply:
A unique derivation string could be used instead of bare sha256(key) to decrease the chances that some other protocol uses sha256(key) for something else

Reply:
@mlsteele that would be nice, but it needs to be something that's consistent and deterministically generated between devices.  Would you suggest making another key type in the implicit team sigchain for the TLF?

Reply:
I'm not suggesting using additional team mechanisms.

[Here](https://github.com/keybase/client/blob/333ded3a693b5acff16faf393dc6e5df0091bd70/go/stellar/note.go#L37) in stellar for example [DeriveSymmetricKey](https://github.com/keybase/client/blob/333ded3a693b5acff16faf393dc6e5df0091bd70/go/libkb/naclwrap.go#L636) is used.

Reply:
@mlsteele I see, thanks!  So basically hmac the key (or the hash of the key) with some hardcoded, app-wide-unique constant.  Good idea, will do that when I code that bit, thanks!

Reply:
yeah i agree with @mlsteele here. i think the approach is fine but if this is an important key that can be used in other contexts, i would recommend a prefix that won't be used in that other context.

Reply:
@maxtaco 	yep, already done with a unique prefix.